### DISCUSSION TITLE
For X509 v3 Cert, How to use API to Add SKID & AKID Extensions?

### DISCUSSION THREAD

**Discussion by jmcpheters** [openssl/openssl#25264] [Q&A] [answered]
I'm working on migrating an application to Openssl 3.x involving two way authentication. The client certs, which are self signed, are created in the migration code as v3. I'm attempting to include recommended extensions added to the certificates, and from what I've read having Subject Key Identifier (SKID) and Authority Key Identiifer (AKID) the same is recommended for self signed certificates.

I can generated a self signed client certificate with the following openssl command:
```
$ OPENSSL_CONF=/home/pmlinux64/scratch/openssl.cnf openssl req -nodes -newkey ed25519 -x509 \
    -days 3650 -keyout ED25519_CMD.key -out ED25519_CMD.cert -new \
    -subj "/C=US/CN=my.client.name/" -addext "extendedKeyUsage=clientAuth" \
    -addext "subjectKeyIdentifier=hash" -addext "authorityKeyIdentifier="keyid:always"
```
Note that I set OPENSSL_CONF to point to an empty openssl.cnf file because I want to be sure all inputs involved in generating the cert are coming from the command line. The cert contains expected v3 extensions:
```
$ openssl x509 -noout -text -in ED25519_CMD.cert | grep -A6 "X509v3 extensions:"
        X509v3 extensions:
            X509v3 Extended Key Usage:
                TLS Web Client Authentication
            X509v3 Subject Key Identifier:
                D3:2A:6B:5F:0F:3C:75:ED:03:67:0D:65:90:F5:41:4A:47:5E:73:EB
            X509v3 Authority Key Identifier:
                keyid:D3:2A:6B:5F:0F:3C:75:ED:03:67:0D:65:90:F5:41:4A:47:5E:73:EB
```
Below as an excerpt of a program that attempts to generate an equivalent cert through the C API.
```
#include <openssl/ssl.h>
#include <openssl/evp.h>
#include <openssl/x509v3.h>
#include <openssl/bio.h>

int main(int argc, char *pArgv[])
{
	EVP_PKEY_CTX 	*pctx = NULL;
	EVP_PKEY 		*key = NULL;
	X509			*cert = X509_new();
	X509_NAME		*Name = NULL;
	X509_EXTENSION	*ext = NULL;
	EVP_MD_CTX		*mctx = NULL;
	BIO 			*fbio = BIO_new_file("ED25519_API.cert", "wb");

	OpenSSL_add_all_algorithms();
	SSL_library_init();

	// set stuff in all certs
	X509_set_version(cert, X509_VERSION_3);
	ASN1_INTEGER_set(X509_get_serialNumber(cert), 0);		// dont care, set to zero
	ASN1_UTCTIME_set(X509_get_notBefore(cert), 0);			// don't care, use 1970
	ASN1_UTCTIME_set(X509_get_notAfter(cert), 2335219200);	// don't care, put far in future (year 2044)

	// set subject and issuer
	Name = X509_get_subject_name(cert);
	X509_NAME_add_entry_by_txt(Name, "C", MBSTRING_ASC, (const uint8_t*)"US", -1, -1, 0);
	X509_NAME_add_entry_by_txt(Name, "CN", MBSTRING_ASC, "www.myclient.com", -1, -1, 0);
	X509_set_issuer_name(cert, Name);

	// generate key and use for cert
	pctx = EVP_PKEY_CTX_new_from_name(NULL, "ED25519", NULL);
	EVP_PKEY_keygen_init(pctx);
	EVP_PKEY_keygen(pctx, &key); 
	X509_set_pubkey(cert, key);

	// example add extension that works
	ext = X509V3_EXT_conf_nid(NULL, NULL, NID_ext_key_usage, "clientAuth");
	X509_add_ext(cert, ext, -1);
	X509_EXTENSION_free(ext);

	// start attempt to add SKID and AKID
	ext = X509V3_EXT_conf_nid(NULL, NULL, NID_subject_key_identifier, "hash");
	X509_add_ext(cert, ext, -1);
	X509_EXTENSION_free(ext);

	ext = X509V3_EXT_conf_nid(NULL, NULL, NID_authority_key_identifier, "keyid:always");
	X509_add_ext(cert, ext, -1);
	X509_EXTENSION_free(ext);
	// end attempt to add SKID and AKID

	// sign cert
	mctx = EVP_MD_CTX_new();
	EVP_DigestSignInit_ex(mctx, NULL, NULL, NULL, NULL, key, NULL);
	X509_sign_ctx(cert, mctx);

	// output cert
	PEM_write_bio_X509(fbio, cert);
	BIO_free_all(fbio);	// TODO more stuff to free...
	return 0;
}
```
Unlike the cert generated by the command line, it does not contain SKID and AKID:
```
$ openssl x509 -noout -text -in ED25519_API.cert | grep -A6 "X509v3 extensions:"
        X509v3 extensions:
            X509v3 Extended Key Usage:
                TLS Web Client Authentication
    Signature Algorithm: ED25519
         99:11:5d:84:74:05:17:04:2c:33:9c:aa:ca:8d:f4:e9:9e:5b:
         e6:fd:a6:8a:84:d0:c5:e9:78:85:28:a9:2a:57:db:2f:9b:2a:
         2a:51:24:9f:fe:ed:d1:03:d4:85:0c:23:8e:f5:f9:81:1f:12:
```
I tried variations, such as creating a X509V3_CTX, CONF and other structures and passing them into alternative functions, like what I saw from the walking through the 'openssl req ...' command line code, but I did not resolve the issue because the only change was output of obviously incorrectly set values:
```
        X509v3 extensions:
            X509v3 Subject Key Identifier:

            X509v3 Authority Key Identifier:
                0.
            X509v3 Extended Key Usage:
                TLS Web Client Authentication
```
In the above program excerpt, the addition of extensions is done after all required information is set (Issuer, Subject, private key),
The actual cert generated has the expected Issuer and Subject for both the command line and API generated certs:
```
$ openssl x509 -noout -text -in ED25519_CMD.cert | grep -e "Issuer:" -e "Subject:"
        Issuer: C = US, CN = www.myclient.com
        Subject: C = US, CN = www.myclient.com
$ openssl x509 -noout -text -in ED25519_API.cert | grep -e "Issuer:" -e "Subject:"
        Issuer: C = US, CN = www.myclient.com
        Subject: C = US, CN = www.myclient.com
```
I could spend more time trying to figure out how the 'openssl req' works, but it's taking me too long and connections to the corresponding servers are successful with certs that do not contain these extensions. Thus I don't want to spend anymore time on something that is a recommendation, but not necessary for my application. Originally I was experimenting with extensions in an effort to troubleshoot connectivity issues, but those are resolved and had nothing to do with the certs, but rather dealt with server re-configurations. So at this point I'm simply curious as to why I could not get generate the correct SKID and AKID for the code snippet.

Another consideration is I'd like the cert generation code to be as simple as possible, a la the KISS principle. It is not worth adding a bunch of complicated code for something not required by the application. I'm hoping the errors in the above program excerpt is obvious to someone more experienced with OpenSSL, and that a simple fix exists. 

If anyone can point out a simple way to overcome the lack of SKID and AKID, that would be appreciated.

**Comment by kulkarniamit** [comment, marked as answer]
After you set the public key for the cert using `X509_set_pubkey()`, you can try something like this to add SKID:
> NOTE: Error checking is not added for brevity
```c
  /* Initialize 'cert' and set public key here */
  X509_EXTENSION *ext = NULL;
  ASN1_OCTET_STRING *skid = NULL;
  unsigned char md[EVP_MAX_MD_SIZE];
  unsigned int md_len = 0;

  X509_pubkey_digest(cert, EVP_sha1(), md, &md_len);
  skid = ASN1_OCTET_STRING_new();
  ASN1_OCTET_STRING_set(skid, md, md_len);
  ext = X509V3_EXT_i2d(NID_subject_key_identifier, 0, skid);
  X509_add_ext(cert, ext, -1);
  X509_EXTENSION_free(ext);
  ASN1_OCTET_STRING_free(skid);
```

To add AKID, here's a sample assuming its a self signed certificate
```c
  X509 *issuer_cert = cert;
  AUTHORITY_KEYID *akid = AUTHORITY_KEYID_new();;
  X509_pubkey_digest(issuer_cert, EVP_sha1(), md, &md_len);
  akid->keyid = ASN1_OCTET_STRING_new();
  ASN1_OCTET_STRING_set(akid->keyid, md, md_len);
  ext = X509V3_EXT_i2d(NID_authority_key_identifier, 0, akid);
  X509_add_ext(cert, ext, -1);
  AUTHORITY_KEYID_free(akid);
```

**Reply by jmcpheters** [reply]
This works. I was under the impression from the OpenSSL docs that if SKID is set to "hash" and AKID to "keyid:always", that during signing the pubkey hash would be assigned to the SKID and then the SKID would be assigned to the AKID. This works on the command line. Apparently there is more involved when using the API.
